python/cudf/cudf/core/column/datetime.py

```diff
@@ -540,65 +526,56 @@ def isocalendar(self) -> dict[str, ColumnBase]:
-        elif isinstance(other, np.timedelta64):
-            other_time_unit = np.datetime_data(other.dtype)[0]
-            if np.isnat(other):
-                return cudf.Scalar(
-                    None,
-                    dtype="timedelta64[ns]"
-                    if other_time_unit not in {"s", "ms", "ns", "us"}
-                    else other.dtype,
-                )
-
-            if other_time_unit not in {"s", "ms", "ns", "us"}:
-                other = other.astype(
-                    np.dtype(f"timedelta64[{self.time_unit}]")
-                )
-
-            return cudf.Scalar(other)
-        elif isinstance(other, str):
-            try:
-                return cudf.Scalar(other, dtype=self.dtype)
-            except ValueError:
-                pass
-
-        return NotImplemented
+        elif is_scalar(other) or (
+            isinstance(other, (cp.ndarray, np.ndarray)) and other.ndim == 0
+        ):
+            if isinstance(other, (cp.ndarray, np.ndarray)) and other.ndim == 0:
+                other = other[()]
+            if is_na_like(other):
+                return super()._normalize_binop_operand(other)
+            elif isinstance(other, pd.Timestamp):
+                if other.tz is not None:
+                    raise NotImplementedError(
+                        "Binary operations with timezone aware operands is not supported."
+                    )
+                other = other.to_numpy()
+            elif isinstance(other, str):
+                try:
+                    other = pd.Timestamp(other)
+                except ValueError:
+                    return NotImplemented
+            elif isinstance(other, (np.datetime64, np.timedelta64)):
+                unit = np.datetime_data(other)[0]
+                if unit not in {"s", "ms", "us", "ns"}:
+                    if np.isnat(other):
+                        # TODO: Use self.time_unit to not modify the result resolution?
+                        to_unit = "ns"
+                    else:
+                        to_unit = self.time_unit
+                    if np.isnat(other):
+                        # Workaround for https://github.com/numpy/numpy/issues/28496
+                        # Once fixed, can always use the astype below
+                        other = type(other)("NaT", to_unit)
+                    else:
+                        other = other.astype(
+                            np.dtype(f"{other.dtype.kind}8[{to_unit}]")
+                        )
+            scalar = pa.scalar(other)
+            if pa.types.is_timestamp(scalar.type):
+                if scalar.type.tz is not None:
+                    raise NotImplementedError(
+                        "Binary operations with timezone aware operands is not supported."
+                    )
+                return scalar
+            elif pa.types.is_duration(scalar.type):
+                return scalar
+            else:
+                return NotImplemented
```

Comment:
Returning NotImplemented is already the fallback in the elif so I think my suggestion from above applies.